sdks/python/apache_beam/runners/worker/bundle_processor.py

```diff
@@ -379,6 +379,39 @@ def _commit(self):
       self._state_handler.blocking_append(self._state_key, out.get())
 
 
+# TODO(BEAM-5428): Implement cross-bundle state caching.
+class SynchronousSetRuntimeState(userstate.RuntimeState):
+  def __init__(self, state_handler, state_key, value_coder):
+    self._state_handler = state_handler
+    self._state_key = state_key
+    self._value_coder = value_coder
+    self._cleared = False
+    self._added_elements = set()
+
+  def read(self):
+    return _ConcatIterable(
+        {} if self._cleared else _StateBackedIterable(
+            self._state_handler, self._state_key, self._value_coder),
+        self._added_elements)
+
+  def add(self, value):
+    self._added_elements.add(value)
+
+  def clear(self):
+    self._cleared = True
+    self._added_elements = set()
+
+  def _commit(self):
+    if self._cleared:
+      self._state_handler.blocking_clear(self._state_key)
+    if self._added_elements:
```

Comment:
One downside of this way of doing things is that if we add the same element many times from different bundles, the underlying bag state keeps growing. Perhaps we should (periodically?) "compact" by reading everything in, eeduplicating via set(), and then writing it all back out. (A TODO is fine here if you don't want to implement it now.)

Reply:
I would prefer to have the compaction logic but I will do some research on my end before making the final decision.

Reply:
@robertwb Check the latest implementation of `read()` method. I think this is what you recommended. Let me know if you want me to implement it in different way.

Reply:
Looks good, thanks.